### PR TITLE
Add a compatibility type for C++20's std::type_identity.

### PR DESCRIPTION
**Pull request by bangerth**
We have long had a `identity` type. In C++20, there is finally `std::type_identity` we can built on instead. Import that into `namespace std_cxx20` and use it to define `identity`.

In a follow-up patch, I will switch all uses of `identity<T>` to `std_cxx20::type_identity<T>`. (Or, rather, to `std_cxx20::type_identity_t<T>`.

### CODE REVIEW

**Review by masterleinad — COMMENTED** [review]
I think we should again use a feature test macro in the `C++20` check we run while configuring via `CMake`.

**Comment by bangerth** [comment]
@masterleinad Maybe like the last commit here?

I think a good question is what we will do when we actually require C++20. Will we continue to check for the existence of features with these feature test macros? What if they are not satisfied? This is a question separate from the current patch, but one I'd be curious to hear how you think about for the long run.

**Comment by masterleinad** [comment]
> @masterleinad Maybe like the last commit here?

We discussed some time ago (https://github.com/dealii/dealii/issues/10340), that it's not a good idea, in general, to use the feature macros directly in our source code but rather while configuring using `CMake`.

> I think a good question is what we will do when we actually require C++20. Will we continue to check for the existence of features with these feature test macros? What if they are not satisfied? This is a question separate from the current patch, but one I'd be curious to hear how you think about for the long run.

We still have checks for `C++14` support and I would think that we will continue to keep those for the oldest version supported. Of course, this also heavily depends on our decision for the oldest compilers we promise to support. If all of those support a particular feature anyway, we don't have to test for it (assuming that newer compilers will not drop support), though.

**Comment by bangerth** [comment]
Ah, thanks for the reminder. How about this commit?

As for the general question: I see these checks as things that are hard to get rid of. If we still have checks for specific C++14 features, then that's likely not because we need them, but because nobody has removed them. But that is not a discussion we need to have here.

**Review by masterleinad — APPROVED** [review]

File: include/deal.II/base/template_constraints.h [hunk 1]
```diff
@@ -333,70 +334,15 @@ constexpr bool has_begin_and_end =
...
- * @endcode
+ * @deprecated Use `std_cxx20::identity_type` instead.
  */
 template <typename T>
-struct identity
-{
-  using type = T;
-};
-
+using identity = std_cxx20::type_identity<T>;
```
Comment (masterleinad) [inline include/deal.II/base/template_constraints.h:345]: We should deprecate this properly when we change all uses in the library.

Comment (bangerth) [reply include/deal.II/base/template_constraints.h:345]: Will do in the follow-up!

**Pull request merged**